File: plugins/plugin-telegram/typescript/src/providers/chatState.ts
import type { IAgentRuntime, Memory, Provider, ProviderResult, State } from "@elizaos/core";

export const CHAT_STATE_PROVIDER = "telegram_chat_state";

export const chatStateProvider: Provider = {
  name: CHAT_STATE_PROVIDER,
  description: "Provides Telegram chat context including chat ID, user ID, and chat type",
  dynamic: true,

  get: async (_runtime: IAgentRuntime, message: Memory, _state: State): Promise<ProviderResult> => {
    const chatId = message.content?.chatId as number | undefined;
    const userId = message.content?.userId as number | undefined;
    const threadId = message.content?.threadId as number | undefined;
    const roomId = message.roomId;

    const isPrivate = chatId !== undefined && chatId > 0;
    const isGroup = chatId !== undefined && chatId < 0;

    const data = {
      chatId,
      userId,
      threadId,
      roomId,
      isPrivate,
      isGroup,
    };

    const values: Record<string, string> = {
      chat_id: chatId?.toString() || "",
      user_id: userId?.toString() || "",
      thread_id: threadId?.toString() || "",
      room_id: roomId || "",
      is_private: isPrivate.toString(),
      is_group: isGroup.toString(),
    };

    let text = "Telegram Chat State:\n";
    if (chatId) {
      text += `Chat ID: ${chatId}\n`;
      text += `Chat Type: ${isPrivate ? "Private" : "Group"}\n`;
    }
    if (userId) {
      text += `User ID: ${userId}\n`;
    }
    if (threadId) {
      text += `Thread ID: ${threadId}\n`;
    }

    return {
      data,
      values,
      text,
    };
  },
};

File: plugins/plugin-telegram/rust/src/providers/chat_state.rs
use async_trait::async_trait;
use serde_json::Value;

use super::{ProviderContext, TelegramProvider};

/// Provider that exposes the current chat/user/thread context as JSON.
pub struct ChatStateProvider;

#[async_trait]
impl TelegramProvider for ChatStateProvider {
    fn name(&self) -> &'static str {
        "telegram_chat_state"
    }

    async fn get(&self, context: &ProviderContext) -> Value {
        serde_json::json!({
            "chat_id": context.chat_id,
            "user_id": context.user_id,
            "thread_id": context.thread_id,
            "room_id": context.room_id,
            "is_private": context.chat_id.map(|id| id > 0).unwrap_or(false),
            "is_group": context.chat_id.map(|id| id < 0).unwrap_or(false)
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_chat_state_private() {
        let provider = ChatStateProvider;
        let context = ProviderContext {
            chat_id: Some(12345),
            user_id: Some(12345),
            thread_id: None,
            room_id: Some("room-uuid".to_string()),
        };

        let state = provider.get(&context).await;
        assert_eq!(state["chat_id"], 12345);
        assert_eq!(state["is_private"], true);
        assert_eq!(state["is_group"], false);
    }

    #[tokio::test]
    async fn test_chat_state_group() {
        let provider = ChatStateProvider;
        let context = ProviderContext {
            chat_id: Some(-12345),
            user_id: Some(67890),
            thread_id: Some(1),
            room_id: Some("room-uuid".to_string()),
        };

        let state = provider.get(&context).await;
        assert_eq!(state["chat_id"], -12345);
        assert_eq!(state["is_private"], false);
        assert_eq!(state["is_group"], true);
    }
}
